### PR TITLE
New Serverless Pattern - REST API via API Gateway, acting as an S3 proxy (SLS & Python)

### PR DESCRIPTION
**Pull request by bb245**
*Issue #, if available:* https://github.com/aws-samples/serverless-patterns/issues/2095

*Description of changes:*

This setup orchestrates the deployment of a REST API via API Gateway, acting as an S3 proxy for write operations. It includes API-Key Setup, basic Request Validator, and Lambda invocation on S3 events. It enables CloudWatch logs and X-Ray tracing for API Gateway and Lambda using AWS Powertool for Lambda framework.

### CODE REVIEW

File: apigw-s3-lambda-sls-py/README.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
+# REST API as an Amazon S3 proxy in API Gateway
+
+This pattern utilizes Serverless framework to deploy a REST API via API Gateway, acting as an S3 proxy for write operations.
```
Comment (bfreiberg) [inline apigw-s3-lambda-sls-py/README.md:3]:
```suggestion
This pattern utilizes the Serverless framework to deploy a REST API via API Gateway, acting as an S3 proxy for write operations. 
```

File: apigw-s3-lambda-sls-py/README.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
+# REST API as an Amazon S3 proxy in API Gateway
+
+This pattern utilizes Serverless framework to deploy a REST API via API Gateway, acting as an S3 proxy for write operations. 
+It includes API-Key Setup, Request Body Validator, Lambda invocation on S3 events, and enables CloudWatch logs and X-Ray tracing for API Gateway and Lambda using AWS Powertool for Lambda framework.
```
Comment (bfreiberg) [inline apigw-s3-lambda-sls-py/README.md:4]:
```suggestion
It includes API-Key Setup, Request Body Validator, Lambda function invocation on S3 events, and enables CloudWatch logs and X-Ray tracing for API Gateway and Lambda using the Powertools for AWS Lambda.
```

Comment (bfreiberg) [reply apigw-s3-lambda-sls-py/README.md:4]: Please include a link to Powertools

File: apigw-s3-lambda-sls-py/README.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    ```commandline
+    sls plugin install -n serverless-python-requirements
+    ```
+   - `serverless-python-requirements` plugin will automatically detect `Pipfile` and pipenv installed and generate a `requirements.txt` file
+
+5. Deploy the services specified in [serverless.yml](./serverless-compose.yml) by running:
+   ```commandline
+      sls deploy --stage dev --verbose
+   ```
+6. The outputs from the Serverless deployment process containing the resource names, IDs and/or ARNs which are used for
```
Comment (bfreiberg) [inline apigw-s3-lambda-sls-py/README.md:48]:
```suggestion
6. The outputs from the Serverless deployment process contain the resource names, IDs and/or ARNs which are used for
```

File: apigw-s3-lambda-sls-py/README.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+   testing.
+
+
+## How it works
+
+The high-level diagram below serves to visually represent this pattern.
+
+![apigw-s3-lambda-sls-py.png](./docs/apigw-s3-lambda-sls-py.png)
+
+This setup involves creating and launching a REST API using API Gateway, allowing direct S3 writing without additional services.
```
Comment (bfreiberg) [inline apigw-s3-lambda-sls-py/README.md:58]:
```suggestion
This setup involves creating and launching a REST API using API Gateway, allowing direct writing to an Amazon S3 bucket without additional services. 
```

File: apigw-s3-lambda-sls-py/README.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+
+## How it works
+
+The high-level diagram below serves to visually represent this pattern.
+
+![apigw-s3-lambda-sls-py.png](./docs/apigw-s3-lambda-sls-py.png)
+
+This setup involves creating and launching a REST API using API Gateway, allowing direct S3 writing without additional services. 
+It utilizes an open-source Serverless framework for deployment. The pattern includes setting up API keys with Usage Plan, Request Body Validator, and triggering Lambda functions on S3 events.
```
Comment (bfreiberg) [inline apigw-s3-lambda-sls-py/README.md:59]:
```suggestion
It utilizes the open-source Serverless framework for deployment. The pattern includes setting up API keys with Usage Plan, Request Body Validator, and triggering Lambda functions on S3 events.
```

File: apigw-s3-lambda-sls-py/README.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+Additionally, this setup establishes Cloudwatch logs and X-Ray tracing for both API Gateway and Lambda, harnessing the capabilities of AWS Powertool specifically for Lambda functions.
```
Comment (bfreiberg) [inline apigw-s3-lambda-sls-py/README.md:60]:
```suggestion
Additionally, this setup establishes Cloudwatch logs and X-Ray tracing for both API Gateway and Lambda, harnessing the capabilities of Powertools specifically for Lambda functions.
```

File: apigw-s3-lambda-sls-py/README.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+#### I. Happy Path
+
+This setup can be tested using the [curl command](https://github.com/curl/curl/blob/master/docs/MANUAL.md) or  [xh utility](https://github.com/ducaale/xh)  by accessing the `PUT /dev/api/{order_object_path+}
+` endpoint. 
+
+In this context, the inclusion of 'suffix' alongside 'order_object_path' suggests that we have the flexibility to substitute multiple folders in place of 'order_object_path'.
+
+```commandline
+xh put https://${ApiGatewayRestApi}.execute-api.${AWS::Region}.amazonaws.com/${sls:stage}/api/orders_recevied/order001.json X-Api-Key:ttLMU7hv2S4qEpBVaCYXt4OWoy38jRmF7txgEBXy orderDate=2024-02-06 orderPaymentAmount:=23.99 orderPaymentCurrency=USD customerId:=10002
```
Comment (bfreiberg) [inline apigw-s3-lambda-sls-py/README.md:72]: Replace the API Key with a placeholer

File: apigw-s3-lambda-sls-py/README.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+x-amzn-requestid: 1dd4f2f1-f5ac-46d6-8f65-0d60d4370f0b
+x-amzn-trace-id: Root=1-65c2e2dd-737319bc469770c954e78a1c
+
+```
+
+#### II. Testing Request Validator:
+In this scenario, a value of "ZZZ" is passed for `orderPaymentCurrency` in the request body, which is not among the permitted values.
+
+```commandline
+xh put https://${ApiGatewayRestApi}.execute-api.${AWS::Region}.amazonaws.com/${sls:stage}/api/orders_recevied/order001.json X-Api-Key:ttLMU7hv2S4qEpBVaCYXt4OWoy38jRmF7txgEBXy orderDate=2024-02-06 orderPaymentAmount:=23.99 orderPaymentCurrency=ZZZ customerId:=10002
```
Comment (bfreiberg) [inline apigw-s3-lambda-sls-py/README.md:105]: Replace API Key with a placeholder

**Comment by bb245** [comment]
@bfreiberg, Thank you for your feedback. All comments in the review have been addressed.

**Comment by bfreiberg** [comment]
Looks good, thanks for your contribution.

**Pull request merged**